### ISSUE TITLE
scope issues in comprehensions

### ISSUE THREAD

**Issue by sqwishy** [kellyjonbrazil/jello#46]
Example JSON

``` json
{
   "foods": [
      { "name": "carrot" },
      { "name": "banana" }
   ],
   "people": [
      { "name": "alice", "likes": "apples" },
      { "name": "bob", "likes": "banana" },
      { "name": "carrol", "likes": "carrot" },
      { "name": "dave", "likes": "donuts" }
   ]
}
```

What happens as of 99f6771

``` shell
$ jello "foods = set(f.name for f in _.foods)
         [p.name for p in _.people if p.likes not in foods]"
jello:  Query Exception:  NameError
        name 'foods' is not defined
        query:  foods = set(f.name for f in _.foods)\n[p.name for p in _.people if p.likes not in foods]
        data:  {'foods': [{'name': 'carrot'}, {'name': 'banana'}], 'people': [{'name': 'alice', 'likes': 'apples'}, {'name': 'bob', 'likes': 'banana'}, {'name': 'carrol', 'likes': 'carrot'}, {'name':
            'dave', 'likes': 'donuts'}]}

$ jello "[p.name for p in _.people if p.likes not in (f.name for f in _.foods)]"
jello:  Query Exception:  NameError
        name '_' is not defined
        query:  [p.name for p in _.people if p.likes not in (f.name for f in _.foods)]
        data:  {'foods': [{'name': 'carrot'}, {'name': 'banana'}], 'people': [{'name': 'alice', 'likes': 'apples'}, {'name': 'bob', 'likes': 'banana'}, {'name': 'carrol', 'likes': 'carrot'}, {'name':
            'dave', 'likes': 'donuts'}]}

```

What I expected to happen

``` shell
$ jello "foods = set(f.name for f in _.foods)
         [p.name for p in _.people if p.likes not in foods]"
[
  "alice",
  "dave"
]

$ jello "[p.name for p in _.people if p.likes not in (f.name for f in _.foods)]"
[
  "alice",
  "dave"
]
```

Suggested change

``` diff
diff --git a/jello/lib.py b/jello/lib.py
index 60b442c..ed2f4cc 100644
--- a/jello/lib.py
+++ b/jello/lib.py
@@ -459,10 +459,12 @@ def pyquery(_θ_data, _θ_query):
     del _θ_query
     _θ_last = ast.Expression(_θ_block.body.pop().value)    # assumes last node is an expression
 
-    exec(compile(_θ_block, '<string>', mode='exec'))
+    _θ_scope = {'_': _}
+
+    exec(compile(_θ_block, '<string>', mode='exec'), _θ_scope)
     del _θ_block
 
-    _θ_output = eval(compile(_θ_last, '<string>', mode='eval'))
+    _θ_output = eval(compile(_θ_last, '<string>', mode='eval'), _θ_scope)
     del _θ_last
 
     # convert output back to normal dict
```

**Comment by kellyjonbrazil** [comment]
Hi there, thanks for reporting this!

Your solution seems to work and tests are passing, so it doesn't look like it breaks anything. Nice!

For my benefit, could you help explain what this fix does? I had not messed around with the `exec` and `eval` scopes as I'm not very familiar with their functionality and I don't want to break anything. Is this basically adding the `_` object to the scope, or is it deleting everything from the scope except for `_`? How did this end up fixing the issue?

I feel like I have run into other oddities in the past and this might be the fix for those as well.

**Comment by sqwishy** [comment]
> How did this end up fixing the issue?

I don't really know why the list comprehension behaves that way. But passing in the scope explicitly is almost always what you want.

I suspect the reason for the creative variable naming and `del` from scope was to avoid user input from accidentally affecting local variables.  But they still have access to globals.

``` shell
$ echo 1 | jello "list(globals().keys())"
[
  "__name__",
  "__doc__",
  "__package__",
  "__loader__",
  "__spec__",
  "__file__",
  "__cached__",
  "__builtins__",
  "os",
  "sys",
  "ast",
  "json",
  "shutil",
  "TextWrapper",
  "DotMap",
  "highlight",
  "Style",
  "Name",
  "Number",
  "String",
  "Keyword",
  "JsonLexer",
  "JavascriptLexer",
  "Terminal256Formatter",
  "HtmlFormatter",
  "PYGMENTS_INSTALLED",
  "opts",
  "JelloTheme",
  "Schema",
  "Json",
  "load_json",
  "warning_message",
  "pyquery"
]
```

Maybe this is unlikely to cause breakage, but if you pass in the global scope explicitly it'll be used for both locals and globals when the user input that is evaluated, *instead of using the current scope*.  Then it might behave more in the way you want it and you shouldn't have to worry about inadvertent side-effects.

So that's what I would suggest. But sorry I can't explain in great detail why this fixes anything other than what is in the docs. (They do mention something about [not having access to nested scopes](https://docs.python.org/3/library/functions.html#eval) but I don't really know if that's what's going on with the list comprehension.) With stuff like this, I look how it's done in the standard library, like in pdb or [runpy](https://docs.python.org/3/library/runpy.html). If I'm trying to write something that does a similar as those modules I'll look at how they work, like how they use `eval` and `exec` in this case, and look at what they do because I more or less don't really know what I'm doing and mainly try to copy smart people and hope things work.

**Comment by sqwishy** [comment]
oh one real quick bonus meme because you mentioned this

>I feel like I have run into other oddities in the past and this might be the fix for those as well.

so I noticed you *remove* the last statement from the block and `eval` it separately in order to get the result of the expression. alternatively, if the last statement in the block is an expression you could *leave it in and change it* to an assignment statement so it stores to some variable '__final' or something and that will be written to the scope dictionary passed in to `exec` when the block is exectued

But I don't actually know if this is better or not. I couldn't find a situation where this actually improved anything so it maybe doesn't. Just an idea if the current approach doesn't work for some reason in the future or whatever.

**Comment by kellyjonbrazil** [comment]
Nice - thanks! Yes, that was a struggle to find out how to do get that to work similar to a REPL. I'll definitely consider the option you describe.

Yes, I was using the strange variable names and `del` to keep from polluting the scope too much, but also wanted some convenience modules pre-imported like `os.environ`. I'll dig into this a bit more. Thanks again!

**Comment by kellyjonbrazil** [comment]
fixed in v1.5.3 available on Pypi. Binaries coming soon.
